docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -56,113 +97,145 @@ The current multisig public keys use amino serialization to generate the address
-	if len(contents) != 0 {
-		preImage = append(preImage, 0)
-		preImage = append(preImage, contents...)
-	}
-	return sha256.Sum256(preImage)[:20]
+const A_LEN = 32
+
+func Hash(typ string, key []byte) []byte {
+    return hash(hash(typ) + key)[:A_LEN]
 }
 ```
 
-`AddressHash` always take a string `prefix` as a starting point which should represent the
-type of public key (ex. `sr25519`) or module account being used (ex. `staking` or `group`).
-For public keys, the `contents` parameter is used to specify the binary contents of the public
-key. For module accounts, `contents` can be left empty (for modules which don't manage "sub-accounts"),
-or can be some module-specific content to specify different pools (ex. `bonded` or `not-bonded` for `staking`)
-or managed accounts (ex. different accounts managed by the `group` module).
+The `+` is a bytes concatenation, which doesn't use any separator.
 
-In the `preImage`, the byte value `0` is used as the separator between `prefix` and `contents`. This is a logical
-choice given that `0` is an invalid value for a string character and is commonly used as a null terminator.
+This algorithm is an outcome from a consulting session with a cryptographer.
+Motivation: this algorithm keeps the address relatively small (length of the `typ` doesn't impact on the length of the final address)
+and it's more secure than [post-hash-prefix-proposal] (which uses first 20 bytes of a pubkey hash, that significantly reduce the address space).
+Moreover the cryptographer motivated the choice to add `typ` in the hash to protect against switch table attack.
 
-### Canonical Public Key Address Prefixes
+We use `address.Hash` function for generating address for all accounts represented by a single key:
+* simple public keys: `address.Hash(keyType, pubkey)`
++ aggregated keys (eg: BLS): `address.Hash(keyType, aggregatedPubKey)`
++ module accounts: `addoress.Hash("module", moduleID)`
+  For module subaccount, or module specific content we append a specific key to the moduleKey:
+  `address.Hash("module", moduleSubaccountKey)`  (note: `moduleSubaccountKey` already composes the `modleKey`).
 
-All public key types will have a unique protobuf message type such as:
 
-```proto
-package cosmos.crypto.sr25519;
+### Composed Account Addresses
 
-message PubKey {
-  bytes key = 1;
+For simple composed accounts (like new naive multisig), we generalize the `address.Hash`. The address is constructed by recursively creating addresses for the sub accounts, sorting the addresses and composing it into a single address. It ensures that the ordering of keys doesn't impact the created address.
+
+```go
+// We don't need a PubKey interface - we need anything which is addressable.
+type Addressable interface {
+    Address() []byte
+}
+
+func Composed(typ string, subaccounts []Addressable) []byte {
+    addresses = map(subaccounts, \a -> a.Address())
+    addresses = sort(addresses)
+    return address.Hash(typ, addresses[0] + ... + addresses[n])
```

Comment:
```suggestion
    return address.Hash(typ, LengthPrefixAddress(addresses[0]) + ... + LengthPrefixAddress(addresses[n]))
```

Reply:
Without length prefixing each address here you don't know the boundary. The addresses could be 20 or 32 bytes or some other number of users choose. So you can't successfully compose without that. Otherwise, there could be some weird collisions however unlikely.

Reply:
`Composed` is a one-way function - once hashed, we can't go back. We don't need to know boundaries - finding a collision with length prefixing is as hard as finding a collision without them. We can map `LengthPrefixAddress` to all sub addresses, but I don't see any justification for that. Meaning it obfuscates the code but it doesn't enhance the security.

Reply:
So, practically speaking, the only way we could get a collision is when we have to sequence of addresses which concatenate to the same string:
```
as, bs [][]byte <- list of addresses
```
Question: is it harder to find 2 different list of addresses (as != bs) for having a match in A or in B?
```
A:  as[0]+as[1]+...+as[n] =? bs[0]+bs[1]+...+bs[m]
B:  lenpref(as[0])+lenpref(as[1])+...+lenpref(as[n]) =? lenpref(bs[0])+lenpref(bs[1])+...+lenpref(bs[m])
```
so, let think...

Reply:
Each problem can be solved by a construction: we map and concatenate. Then for constructing `bs` we try to slice the concatenate sequence just before some `1` bits. 
Specifically, for problem `B`, here is an example of conflicting addresses in binary form:
```
as = {0101, 1000, 1}
bs = {00, 10 1000  11}
map(lenpref, as) = {10 0  100,  10 1000, 11  }
map(lenpref, as) = {10 0, 100   10 1000  11 }
```

Reply:
I was thinking more about it. Since the length prefix is always 1 byte, then we can't construct conflicting address sequences unless the individual addresses are the same - so the `lenpref` in this case will work. 
Other solution I had in mind was to hash each subaddress before concatenation - this way all parts have exactly same length.

Reply:
Yeah, the collision risk may be small but length-prefixing is cheap and ensures zero collisions in the pre-image.

Reply:
@aaronc , yes, in OP, when you wrote about collision I thought you are writing about hash collision ;) In the (updated) text I use _conflict_ instead of _collision_.